### DISCUSSION TITLE
[i18n] Accessing /mypage should redirect to /{locale}/mypage

### DISCUSSION THREAD

**Discussion by romaincointepas** [vercel/next.js#22405] [Ideas]
**Describe the feature you'd like to request**

Right now locale detection only happens at the root (`/` -> `/fr`).

In many cases pages are shared/accessed directly, and in that case `/mypage` should redirect to `/fr/mypage`.

Example of popular websites with this behavior:
- https://www.microsoft.com/windows (-> https://www.microsoft.com/fr-fr/windows)
- https://stripe.com/payments (-> https://stripe.com/fr/payments)

**Describe the solution you'd like**

Apply locale detection and redirect to all paths (root and subpaths).

**Describe alternatives you've considered**

None available.

**Comment by timbset** [comment]
I subscribe to every word been written above. I expect that every path without locale will be redirected to some specific locale. The only way to implement this behaviour is parse `NEXT_LOCALE` cookie and `Accept-Language` header to detect redirect request in `getInitialProps`. I hope that it will be implement out-of-box some time, maybe with the option in next config.

**Comment by gpaluk** [comment]
I'm here because I raised the same issue as a bug however that bug report was promptly closed as this is currently the expected behaviour. I would like to +1 this as a topic for discussion as it doesn't seem to align with expected behaviour from a user or web crawler perspective.

**Comment by ababol** [comment]
I don't know if I might help someone but I've forked nextjs and add a "hack" on top of their locale detection to handle more pages than just the root one ; but it's only working in SSR.
I end up adding `localeDetectionPaths` on my `next.config.js` to handle which pages should redirect.

I encounter another issue which is: on my project I've the cache activated on my pages, the cache mess up the locale detection redirect.
So I've also added a `catchAll` lang which is the default locale `/`, this allow me to be able to activate cache without losing the ability to redirect.

See my next.config.js:
```
  i18n: {
    locales: ["en", "fr", "es", "pt", "catchAll"],
    defaultLocale: "catchAll"
    localeDetectionPaths: [
      "^/$",
      "^/streams$",
      "^/events/(.*)/shop$",
      "^/events/(.*)/stream$",
      "^/venues/(.*)",
      "^/orders(.*)",
      "^/profile$",
    ],
  }
```

Here is the [fork of nextjs containing those params](https://github.com/shotgun-warehouse/next.js/commit/659abc76c74adbdce508cd56359d1ccde9e2a17b).

You can try the solution here: https://staging.shotgun.live/streams

Hope this can help.

**Comment by hohl** [comment]
This really should be the default. Any URL without a locale in the path should auto-detect the browser language and redirect to a URL with the local in the path. And, yes, even if—for example `en` is the default—the default locale should be part of the URL too to not re-trigger the detection.

At the moment the lack of enforcing the default local as part of the URL actually produces non-unique URLs. Right now, there are two URLs for all pages in the default locale (for ex. `/about` and `/en/about`).

The examples by OP actually shows it pretty well already:

`https://www.microsoft.com/windows/` <- auto-detection URL
`https://www.microsoft.com/en-us/windows/` <- English variant
`https://www.microsoft.com/fr-fr/windows/` <- French variant
... and so on

That's the way I've seen it from pretty much all multi-national websites and would say has become the de-facto standard in the web. I don't want my websites to break this standard just because they are relying on NextJS.

**Reply by raimille1** [reply]
I 100% agree with this approach^

**Comment by timscott** [comment]
I agree this should be the default or at least an option. I wonder if it's a matter of implementation difficulty, or does the NextJs team think it's preferable this way?

Does anyone know of any workaround? Obviously, one option is make a Choose Language selector so the user is not stuck, but that's a poor UX.

**Reply by gpaluk** [reply]
We already created a dropdown selector however we consider this a separate issue for the following reasons:
a) SEO is still affected
b) It's a patch rather than a fix and doesn't provide the same UX

**Comment by raimille1** [comment]
I've been thinking on how to do this also for kalendme.com ... 

### **Context (skip to implementation if not intersted)**
We're a trtilingual scheduling platform focused on LATAM and the US ([kalendme.com](kalendme.com)). The idea is to give users a link so they can share their availability (ala calendly, but much cheaper and better of course ;)), but the point is that we want to make cross country sharing seamless as we've seen with our users and our competitors, no one does this perfectly. 

So when I send my link to someone else, I want that someone else to feel it was made for them, so it speaks their language automatically and it is in their time zone without anyone configuring anything.

### **Our implementation**
Unfortunately, we had to give away static site optimization by adding getInitialPropos to _app to make this work. There's two reasons:
1) The one mentioned in the main topic of the discussion, most of our users come in through a link that is not / ... so using the default language ('pt') is not an option.
2) Next.JS seem to be using @hapi/accept for parsing the Accept-Language header, but they only support strict accept-language headers and there's browsers out there (including Chrome for iOS) that don't send it that way. Chrome for iOS sends the Accept-Language header with just the locale like this: 'en-US' ... so there's no negotiation that can happen. 
3) I really don't want to fork Next ....

Mixing those two issues, we ended up implementing this:
`_app.js`

```
...
KalendMeApp.getInitialProps = async (appContext) => {
  const { ctx: { req, res }, ctx } = appContext
  const { userLanguage, shouldRedirectLanguage } = configureUserLanguage(ctx)
  if (shouldRedirectLanguage) {
    if (req) {
      res.writeHead(301, { Location: `/${userLanguage}${req.url}` })
      res.end()
      return { }
    } else {
      Router.push(`/${userLanguage}${req.url}`)
    }
  }
  return { ...await App.getInitialProps(appContext) }
}
```

`configureUserLanguage`

```
parser: https://github.com/opentable/accept-language-parser

export default function configureUserLanguage(context) {
  const { req, locale } = context
  if (!req) {
    return { userLanguage, shouldRedirectLanguage: false }
  }
  const nextLocale = req.cookies?.NEXT_LOCALE
  let userLanguage
  if (nextLocale) {
    userLanguage = nextLocale
  } else {
    userLanguage = parser.pick(i18n.locales, req.headers["accept-language"], { loose: true }) || i18n.defaultLocale
  }
  const shouldRedirectLanguage = (i18n.defaultLocale === locale && userLanguage !== locale)
  if (shouldRedirectLanguage) {
    context.res.setHeader('set-cookie', `NEXT_LOCALE=${userLanguage};path=/`)
  }
  return { userLanguage, shouldRedirectLanguage }
}
```

### **Takeaway**
This new parser solves issue 2, and issue 1 is solved in the getInitialProps for _app.

I honestly don't love it but seems to be the cleanest solution I've found over the past days, I hope it helps someone looking this up or if anyone has any suggestions/comments on it would love to hear them! We really want to keep static site optimization for all our static pages.

**Reply by JannikSt** [reply]
Cheers @raimille1, 
are you still running the approach in production? 
Did you face any `[ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client` issues during dev mode?

**Reply by raimille1** [reply]
Hi @JannikSt ! Where still using this in prod. Some cases we’ve seen the detection fails, and we do run into the cannot set headers after restarting a dev process intermittently. It hasn’t become annoying enough for us to fix it, but we’re quickly getting there.

 Have you read about any other solutions? I can’t believe we’re so few facing these challenges.

**Comment by daviesdoclc** [comment]
You can get all this to work by exporting 'redirects' from next.config.js
https://nextjs.org/docs/api-reference/next.config.js/redirects
```
    async redirects() {
        const regLanguages = languages.map(it => it.key).join('|')
        return [
            {
                source: '/en/:path*',
                has: [
                    {
                        type: 'cookie',
                        key: 'lang',
                        value: `(?<lang>${regLanguages})`
                    }
                ],
                destination: '/:lang/:path*',
                permanent: false,
                locale: false
            },
            {
                source: '/en/:path*',
                has: [
                    {
                        type: 'header',
                        key: 'accept-language',
                        value: `(?<lang>${regLanguages})-(?<country>.*)`
                    }
                ],
                destination: '/:lang/:path*',
                permanent: false,
                locale: false
            }
        ]
    }

```
I'm still playing around with this, but it seems promising.  I couldn't get the earlier suggestions above to work.

**Reply by jwoo0122** [reply]
Where did you get `languages` variable?

**Comment by imshubhamsingh** [comment]
I think the issue is with the redirect function defined in next.js https://github.com/vercel/next.js/blob/canary/packages/next/shared/lib/i18n/get-locale-metadata.ts#L80-L116. Here it is only checking for root path - `/`. I don't really know the reason for it. Maybe maintainers will be able to give more insights into this. 

cc @timneutkens @leerob

**Comment by leerob** [comment]
I'm going to lock this discussion since I didn't see there was a duplicate. A solution has been posted in [here](https://github.com/vercel/next.js/discussions/18419#discussioncomment-1561577) - let's continue the conversation there!
